Source/Utils/ParticleUtils.H

```diff
@@ -31,6 +32,96 @@ namespace ParticleUtils {
     findParticlesInEachCell( int const lev, amrex::MFIter const& mfi,
                              WarpXParticleContainer::ParticleTileType const& ptile);
 
+    /**
+     * \brief Return (relativistic) particle energy given velocity and mass.
+     * Note the use of `double` since this calculation is prone to error with
+     * single precision.
+     *
+     * @param[in] v2 square of particle speed ( v dot v)
+     * @param[in] mass particle mass
+     * @param[out] energy particle energy in eV
+     */
+    AMREX_GPU_HOST_DEVICE AMREX_INLINE
+    void getEnergy ( amrex::ParticleReal const v2, double const mass,
+                     double& energy )
+    {
+        using std::sqrt;
+        using namespace amrex::literals;
+
+        // Compute c^2 since it is used repeatedly
+        constexpr auto c2 = PhysConst::c * PhysConst::c;
+        energy = mass * c2 * (sqrt(1.0_prt + v2 / c2) - 1.0_prt) / PhysConst::q_e;
+    }
+
+    /**
+     * \brief Return (relativistic) collision energy assuming the target (with
+     * mass M) is stationary and the projectile is approaching with the
+     * the given speed and mass m. Note the use of `double` since this
+     * calculation is prone to error with single precision.
+     *
+     * @param[in] v2 square of particle speed gamma^2 * ( v dot v)
+     * @param[in] m, M mass of projectile and target, respectively
+     * @param[out] energy particle energy in eV
+     */
+    AMREX_GPU_HOST_DEVICE AMREX_INLINE
+    void getCollisionEnergy ( amrex::ParticleReal const v2, double const m,
+                     double const M, double& gamma,
+                     double& energy )
+    {
+        using std::sqrt;
+        using namespace amrex::literals;
+
+        // Compute c^2 since it is used repeatedly
+        constexpr auto c2 = PhysConst::c * PhysConst::c;
+
+        auto const beta2 = v2 / c2;
+        gamma = sqrt(1.0_prt + beta2);
+        auto const m_tot = M + gamma*m;
+        energy = (sqrt(m_tot*m_tot - m*m*beta2) - m - M) * c2 / PhysConst::q_e;
```

Comment:
This expression could probably also be rewritten to be more tolerant to round off fro small v2.

Reply:
Yeap. I can write is as:
```
m_tot = M + m;
energy = 2.0 * m * M * u2 / (gamma + 1) / (sqrt(m_tot*m_tot + 2*m*M*beta2/(gamma + 1)) + m_tot)
```
I'll quickly test to see if the benchmark calculation is successful when using these expressions and single precision particles (reverting to using `ParticleReal` for the energy and mass).

Reply:
I found that the calculation still requires `double`s for the energy and mass. But I did swap the expression for the above one since it is clearer how this reduces to the classical collision energy as gamma -> 1.